File: Programmers/Python/자료구조/과제 진행하기.py
def solution(plans):
    answer = []
    arr = []
    for plan in plans:
        name = plan[0]
        h, m = map(int, plan[1].split(':'))
        t = int(plan[2])
        arr.append((name, h*60 + m, t))

    arr.sort(key= lambda x : x[1])
    st = []
    for i in range(len(arr)-1):
        n, s, t = arr[i]
        if s + t <= arr[i+1][1]:
            answer.append(n)
            time_left = arr[i+1][1] - s - t # 새로운 과제 끝내고 남은 시간
            while st:
                name, time = st.pop()
                if time_left >= time:
                    time_left -= time
                    answer.append(name)
                else:
                    st.append((name, time - time_left))
                    break

        else:
            st.append((n, t - (arr[i+1][1] - s)))
    answer.append(arr[-1][0])

    while st:
        answer.append(st.pop()[0])
    return answer

File: Programmers/Java/과제진행하기.java
package Programmers.Java;

import java.util.*;

class Solution {
    public String[] solution(String[][] plans) {
        List<String> answer = new ArrayList<>();
        Deque<Plan> stack = new ArrayDeque<>();

        Arrays.sort(plans, (a, b) -> a[1].compareTo(b[1]));


        for (int i = 0; i < plans.length; i++) {
            String[] p = plans[i];
            Plan plan = new Plan(p[0],p[1],p[2]);

            if (!stack.isEmpty()) {
                Plan prevPlan = stack.pop();
                int timeLeft = plan.getStartTime() - prevPlan.getStartTime();
                if (prevPlan.getDuration() > timeLeft) {
                    prevPlan.setDuration(prevPlan.getDuration() - timeLeft);
                    stack.push(prevPlan);
                }
                else {
                    answer.add(prevPlan.getName());
                    timeLeft -= prevPlan.getDuration();
                    while (!stack.isEmpty()) {
                        Plan stoppedPlan = stack.pop();
                        if (timeLeft < stoppedPlan.getDuration()) {
                            stoppedPlan.setDuration(stoppedPlan.getDuration() - timeLeft);
                            stack.push(stoppedPlan);
                            break;
                        } else {
                            answer.add(stoppedPlan.getName());
                            timeLeft -= stoppedPlan.getDuration();
                        }
                    }
                }

            }
            stack.push(plan);

        }

        while (!stack.isEmpty()) {
            answer.add(stack.pop().getName());
        }

        return answer.toArray(String[]::new);
    }

    static class Plan {
        private String name;
        private int startTime;
        private int duration;

        public Plan(String name, String startTime, String duration) {
            this.name = name;
            this.startTime = timeToMinute(startTime);
            this.duration = Integer.parseInt(duration);
        }

        private int timeToMinute(String startTime) {
            String[] hm = startTime.split(":");
            return 60 * Integer.parseInt(hm[0]) + Integer.parseInt(hm[1]);
        }

        public String getName() {
            return this.name;
        }
        public int getStartTime() {
            return this.startTime;
        }
        public int getDuration() {
            return this.duration;
        }

        public void setDuration(int duration) {
            this.duration = duration;
        }

    }

}
